node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+	goheadersync "github.com/celestiaorg/go-header/sync"
+	ds "github.com/ipfs/go-datastore"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"go.uber.org/multierr"
+
+	"github.com/rollkit/rollkit/config"
+	"github.com/rollkit/rollkit/p2p"
+	"github.com/rollkit/rollkit/types"
+)
+
+type BlockExchangeService struct {
+	conf         config.NodeConfig
+	genesis      *tmtypes.GenesisDoc
+	p2p          *p2p.Client
+	ex           *goheaderp2p.Exchange[*types.Block]
+	syncer       *goheadersync.Syncer[*types.Block]
+	sub          *goheaderp2p.Subscriber[*types.Block]
+	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
+
+	logger log.Logger
+	ctx    context.Context
+}
+
+func NewBlockExchangeService(ctx context.Context, store ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, logger log.Logger) (*HeaderExchangeService, error) {
+	// store is TxnDatastore, but we require Batching, hence the type assertion
+	// note, the badger datastore impl that is used in the background implements both
+	storeBatch, ok := store.(ds.Batching)
+	if !ok {
+		return nil, errors.New("failed to access the datastore")
+	}
+	ss, err := goheaderstore.NewStore[*types.SignedHeader](storeBatch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the header store: %w", err)
+	}
+
+	return &HeaderExchangeService{
+		conf:         conf,
+		genesis:      genesis,
+		p2p:          p2p,
+		ctx:          ctx,
+		headerStore:  ss,
+		logger:       logger,
+		syncerStatus: new(SyncerStatus),
+	}, nil
+}
+
+func (bExService *BlockExchangeService) initBlockStoreAndStartSyncer(ctx context.Context, initial *types.Block) error {
+	if err := bExService.blockStore.Init(ctx, initial); err != nil {
+		return err
+	}
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
+	defer bExService.syncerStatus.m.Unlock()
+	bExService.syncerStatus.started = true
+	return nil
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
+}
+
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
```

Comment:
For all these, if there is an error we should return. If we can't init nor start, we can't do anything else, so all those errors would just be spamming the logs.

Reply:
Made it to return the error in case of a failed init or start but we still log errors in case it failed to broadcast. Since this is used in `blockPublishLoop` which just relays blocks, it's fine to log errors then.